### PR TITLE
AE-1263: loan events monitor

### PR DESCRIPTION
**Pull request by arbitrary-sparrow**
This PR creates a bot that monitors for Deposit, Withdraw, Borrow, Repay, and LiquidationCall events.

### CODE REVIEW

File: loan-events/.dockerignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+node_modules
+dist
```
Comment (echoAE) [inline loan-events/.dockerignore:2]:
```suggestion
dist
forta.config.json
```

File: loan-events/src/agent.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  Finding, FindingSeverity, FindingType, ethers, getEthersProvider,
+} = require('forta-agent');
+
+const {
+  getAbi,
+  extractEventArgs,
+  getEventInfo,
+} = require('./utils');
+
+// load any agent configuration parameters
```
Comment (echoAE) [inline loan-events/src/agent.js:11]:
```suggestion
// load any bot configuration parameters
```

File: loan-events/src/agent.spec.js [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+// utility function specific for this test module
+// we are intentionally not using the Forta SDK function due to issues with
+// jest mocking the module and interfering with default function values
+function createTransactionEvent(txObject) {
+  const txEvent = new TransactionEvent(
+    null,
+    null,
+    null,
+    [],
+    [],
```
Comment (echoAE) [inline loan-events/src/agent.spec.js:36]:
```suggestion
    {},
```

File: loan-events/src/agent.spec.js [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+  });
+
+  it('contracts key values must be valid', () => {
+    const { contracts } = config;
+    Object.keys(contracts).forEach((key) => {
+      const { address, abiFile, events } = contracts[key];
+
+      // only check that an address exists for the LendingPoolAddressesProvider entry
+      if (key === 'LendingPoolAddressesProvider') {
+        expect(address).not.toBe(undefined);
```
Comment (echoAE) [inline loan-events/src/agent.spec.js:77]: Any interest in adding a check that the `address` value is a valid address?
```suggestion
        expect(address).not.toBe(undefined);
        
        // check that the address is a valid address
        expect(ethers.utils.isHexString(address, 20)).toBe(true);
```

File: loan-events/src/agent.spec.js [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+        },
+      } = config;
+      const { abiFile, events } = lendingPool;
+      abi = utils.getAbi(abiFile);
+
+      const results = getEventFromConfig(abi, events);
+      eventInConfig = results.eventInConfig;
+      eventNotInConfig = results.eventNotInConfig;
+      findingType = results.findingType;
+      findingSeverity = results.findingSeverity;
```
Comment (echoAE) [inline loan-events/src/agent.spec.js:148]:
```suggestion
      ({
        eventInConfig,
        eventNotInConfig,
        findingType,
        findingSeverity
      } = getEventFromConfig(abi, events));
```

File: loan-events/src/agent.spec.js [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+        abi.push(eventNotInConfig);
+      }
+
+      iface = new ethers.utils.Interface(abi);
+
+      // initialize mock transaction event with default values
+      mockTxEvent = createTransactionEvent({
+        logs: [
+          {
+            name: '',
```
Comment (echoAE) [inline loan-events/src/agent.spec.js:181]:
```suggestion
```

File: loan-events/src/agent.spec.js [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+            address: '',
+            signature: '',
+            topics: [],
+            data: '0x',
+            args: [],
```
Comment (echoAE) [inline loan-events/src/agent.spec.js:186]:
```suggestion
```

File: loan-events/src/agent.spec.js [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+      // encode event data
+      // valid event name with valid name, signature, topic, and args
+      const { mockArgs, mockTopics, data } = createMockEventLogs(eventInConfig, iface);
+
+      // update mock transaction event
+      const [defaultLog] = mockTxEvent.logs;
+      defaultLog.name = contractName;
+      defaultLog.address = ethers.constants.AddressZero;
+      defaultLog.topics = mockTopics;
+      defaultLog.args = mockArgs;
```
Comment (echoAE) [inline loan-events/src/agent.spec.js:207]:
```suggestion
```

File: loan-events/src/agent.spec.js [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+      defaultLog.data = data;
+      defaultLog.signature = iface
+        .getEvent(eventInConfig.name)
+        .format(ethers.utils.FormatTypes.minimal)
+        .substring(6);
```
Comment (echoAE) [inline loan-events/src/agent.spec.js:212]:
```suggestion
```

File: loan-events/src/agent.spec.js [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+      const [defaultLog] = mockTxEvent.logs;
+      defaultLog.name = contractName;
+      defaultLog.address = mockContractAddress;
+      defaultLog.topics = mockTopics;
+      defaultLog.args = mockArgs;
+      defaultLog.data = data;
+      defaultLog.signature = iface
+        .getEvent(eventNotInConfig.name)
+        .format(ethers.utils.FormatTypes.minimal)
+        .substring(6);
```
Comment (echoAE) [inline loan-events/src/agent.spec.js:233]:
```suggestion
      defaultLog.address = mockContractAddress;
      defaultLog.topics = mockTopics;
      defaultLog.data = data;
```

File: loan-events/src/test-utils.js [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+  uint16: 0,
+  'uint16[]': [0],
+  uint256: 0,
+  'uint256[]': [0],
+  address: ethers.constants.AddressZero,
+  'address[]': [ethers.constants.AddressZero],
+  bytes: '0xff',
+  'bytes[]': ['0xff'],
+  bytes32: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,
+  'bytes32[]': [0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF],
```
Comment (echoAE) [inline loan-events/src/test-utils.js:13]: Out of curiosity, should both of these types have default values that are strings?  Conversely, should both of these be integers specified by hex-encoded values?

**Review by echoAE — APPROVED** [review]
Minor suggestions about destructuring and one `agent` -> `bot` swap.

File: loan-events/src/test-utils.js (hunk 11 above)
Comment (arbitrary-sparrow) [reply loan-events/src/test-utils.js:13]: Pretty sure `ethers.js` handles both cases, but it does look like the default value for a bytes-like object in `ethers.js` is a string.

File: loan-events/src/agent.spec.js [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+    const mockContractAddress = ethers.utils.getAddress('0x230E76a625927709618AB27761d574A5CFf75d61');
+    const contractName = 'LendingPool';
+
+    beforeEach(async () => {
+      initializeData = {};
+
+      mockContract.getLendingPool = jest.fn().mockResolvedValue(mockContractAddress);
+
+      // initialize the handler
+      await (provideInitialize(initializeData))();
```
Comment (arbitrary-CodeBeholder) [inline loan-events/src/agent.spec.js:131]:
```suggestion
      await provideInitialize(initializeData)();
```

File: loan-events/src/agent.spec.js [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+      handleTransaction = provideHandleTransaction(initializeData);
+
+      ({ protocolName, protocolAbbreviation, developerAbbreviation } = config);
```
Comment (arbitrary-CodeBeholder) [inline loan-events/src/agent.spec.js:134]: nit, but these parens are superfluous

```suggestion
      { protocolName, protocolAbbreviation, developerAbbreviation } = config;
```

File: loan-events/src/agent.spec.js [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+    it('returns empty findings if contract address does not match', async () => {
+      // encode event data
+      // valid event name with valid name, signature, topic, and args
+      const { mockArgs, mockTopics, data } = createMockEventLogs(eventInConfig, iface);
+
+      // update mock transaction event
+      const [defaultLog] = mockTxEvent.logs;
+      defaultLog.address = ethers.constants.AddressZero;
+      defaultLog.topics = mockTopics;
+      defaultLog.data = data;
```
Comment (arbitrary-CodeBeholder) [inline loan-events/src/agent.spec.js:205]: nitpick, but I feel like it's clearer and cleaner to create the log object and push it to the array here, especially in the cases where a test needs multiple logs

```suggestion
      const loanEvent = {
        address: ethers.constants.AddressZero,
        topics: mockTopics,
        data: data,
      };
      
      mockTxEvent.logs.push(loanEvent);
```

**Review by arbitrary-CodeBeholder — COMMENTED** [review]
nothing major

File: loan-events/.dockerignore [hunk 15]
```diff
@@ -0,0 +1,3 @@
...
+forta.config.json
```
Comment (Nomad-AE) [inline loan-events/.dockerignore:3]:
```suggestion
forta.config.json

```
nit: To be POSIX compliant, end the file with a newline

File: loan-events/.eslintrc.js [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+  overrides: [
+    {
+      files: '*',
+      rules: {
+        'no-plusplus': 'off',
+        'no-continue': 'off',
+      },
+    },
+  ],
+};
```
Comment (Nomad-AE) [inline loan-events/.eslintrc.js:26]:
```suggestion
};

```

File: loan-events/.gitignore [hunk 17]
```diff
@@ -0,0 +1,5 @@
+*.swp
+.env
+forta.config.json
+node_modules
+dist
```
Comment (Nomad-AE) [inline loan-events/.gitignore:5]:
```suggestion
dist

```

File: loan-events/Dockerfile [hunk 18]
```diff
@@ -0,0 +1,10 @@
+FROM node:14.15.5-alpine
+ENV NODE_ENV=production
+WORKDIR /app
+LABEL "network.forta.settings.agent-logs.enable"="true"
+COPY bot-config.json ./
+COPY ./src ./src
+COPY ./abi ./abi
+COPY package*.json ./
+RUN npm ci --production
+CMD [ "npm", "run", "start:prod" ]
```
Comment (Nomad-AE) [inline loan-events/Dockerfile:10]:
```suggestion
CMD [ "npm", "run", "start:prod" ]

```

File: loan-events/abi/LendingPool.json [hunk 19]
```diff
@@ -0,0 +1,1082 @@
...
+          "internalType": "uint256",
+          "name": "",
+          "type": "uint256"
+        }
+      ],
+      "stateMutability": "nonpayable",
+      "type": "function"
+    }
+  ]
+}
```
Comment (Nomad-AE) [inline loan-events/abi/LendingPool.json:1082]:
```suggestion
}

```

File: loan-events/bot-config.json [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+          "severity": "Info"
+        },
+        "LiquidationCall": {
+          "type": "Info",
+          "severity": "Info"
+        }
+      }
+    }
+  }
+}
```
Comment (Nomad-AE) [inline loan-events/bot-config.json:36]:
```suggestion
}

```

File: loan-events/package.json [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+    "forta-agent": "^0.1.3"
+  },
+  "devDependencies": {
+    "eslint": "^7.32.0",
+    "eslint-config-airbnb-base": "^14.2.1",
+    "eslint-plugin-import": "^2.24.2",
+    "jest": "^27.0.6",
+    "nodemon": "^2.0.8"
+  }
+}
```
Comment (Nomad-AE) [inline loan-events/package.json:32]:
```suggestion
}

```

File: loan-events/src/agent.spec.js [hunk 22]
```diff
@@ -0,0 +1,255 @@
...
+    null,
+    txObject.logs,
+    null,
+  );
+  return txEvent;
+}
+
+// check the configuration file to verify the values
+describe('check bot configuration file', () => {
+  it('procotolName key required', () => {
```
Comment (Nomad-AE) [inline loan-events/src/agent.spec.js:46]:
```suggestion
  it('protocolName key required', () => {
```
Copypasta typo. It's everywhere.

File: loan-events/src/agent.spec.js [hunk 23]
```diff
@@ -0,0 +1,255 @@
...
+    const { protocolName } = config;
+    expect(typeof (protocolName)).toBe('string');
```
Comment (Nomad-AE) [inline loan-events/src/agent.spec.js:48]: Nit: typeof is a an operator so the parentheses are not needed. `typeof (protocolName)` => `typeof protocolName`. It's in all of our other code so you can take it or leave it.

**Review by Nomad-AE — APPROVED** [review]
Looks good, just minor newlines needed.

File: loan-events/src/agent.spec.js (hunk 13 above)
Comment (arbitrary-sparrow) [reply loan-events/src/agent.spec.js:133]: These parens are required when destructuring using variables that have already been declared.

**Pull request merged**